Web: Add notification store

kimlisa opened the pull request:
part of https://github.com/gravitational/teleport.e/issues/2242

note: recommend reviewing by commit

- Adds a notification store
   - only supports `access_list` notifications atm (for review due dates), and first iteration, it will be `dumb` in that we just fetch it upon init, and updates whenever a user views, lists, edits access list (this is in another PR)
   - not sure how this will grow in the future because we only have one use case, so i tried to make it kinda generic
- Adds a notification bell on the topbar, when there are notification a `attention dot` renders
- Adds a attention dot to main navigation switcher and on the navigation item

with the brain: 

<img width="300" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/81fa60e9-fa7f-4607-96cc-95aafc703e73">


Can view demos in storybook:

<img width="249" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/c56ebb98-4ad7-4211-b569-e13c4d627959">

<img width="522" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/ab9d4cee-2a9c-4f7b-9200-3a288906d9c1">


<img width="248" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/efef32ee-1f64-4b1e-a379-599372700158">

<img width="524" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/5778d781-6844-4f80-bd70-00a870473f25">

<img width="498" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/0bc8eb7c-13a0-4a02-ae6e-3902213b0414">

<img width="265" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/1963892a-3348-406e-98da-fc7659dd5abb">

kimlisa committed:
Add notification bell to top bar

Also fixes topbar.test act warnings (even though tests passed)

kimlisa committed:
Add attention dot to navigation switcher

kimlisa committed:
Add attention dot to navigation item

kimlisa committed:
Create store notification and enable it

kimlisa commented on web/packages/teleport/src/Navigation/Navigation.tsx:
```diff
@@ -176,7 +176,15 @@ export function Navigation({
           onChange={handleCategoryChange}
           value={view}
-          items={NAVIGATION_CATEGORIES}
+          items={[
+            {
+              category: NavigationCategory.Management,
+              requiresAttention: ctx.storeNotifications
+                .getNotifications()
+                .some(n => n.kind === 'access-lists'),
+            },
+            { category: NavigationCategory.Resources },
+          ]}
```
woops... this should've been in the previous commit

kimlisa commented:
friendly ping @gzdunek @ryanclark

gzdunek commented on web/packages/teleport/src/TopBar/Notifications/Notifications.tsx:
```diff
@@ -0,0 +1,129 @@
+  right: 15px;
+`;
+
+export const NotificationItemButton = styled(DropdownItemButton)`
```
IMO that 0.8 opacity doesn't look good here:
<img width="567" alt="image" src="https://github.com/gravitational/teleport/assets/20583051/c7953338-041e-4380-aec4-a62e4791d558">

It makes the items look disabled.
I'm wondering if we need it at all, we change the background, so it is clear which item is highlighted.

kimlisa replied:
i agree, i removed the opacity

gzdunek commented on web/packages/teleport/src/stores/storeNotifications.ts:
```diff
@@ -0,0 +1,61 @@
+
+type NoticeKinds = 'access-lists';
+
+export type Notice = {
+  kind: NoticeKinds;
+  id: string;
+  resourceName: string;
+  date: Date;
+  route: string;
+};
```
I'm wondering if we could make it more generic. Currently, we are requiring from all notification kinds to have things like `resourceName` and `route`. 
I'm pretty sure that in the future we will have items that are not related a resource or doesn't link to anything.

Maybe we could do something more in this shape:
```ts
type AccessListNotification = {
  kind: 'access-lists';
  resourceName: string;
  route: string;
};

export type Notification = {
  item: AccessListNotification;
  id: string;
  date: Date;
};
```

gzdunek reviewed (commented):
FYI we already have `Notifications` in `shared` (you can find them in storybook). We are using them in Connect and Isaiah used them [for a desktop session warnings](http://localhost:9002/?path=/story/teleport-desktopsession--warnings). I'm wondering if these things could (and should) be merged. 
I also know that our design team is working on a [design system for notifications](https://www.figma.com/file/Gpjs9vjhzUKF1GDbeG9JGE/Application-Design-System?type=design&node-id=8890-5502&mode=design&t=STFnnxzxKYsQPFJ3-0). 

Probably these things are different enough that they should be separated (notifications from `shared` are more like alerts, while your notifications are more like an inbox). I'm only wondering if there is a long term vision in this - for example, you when you get a notification, you first see it as an alert, and then it goes to that _inbox_.

kimlisa commented on web/packages/teleport/src/stores/storeNotifications.ts:
```diff
@@ -0,0 +1,61 @@
+    this.setState({ notices: [...sortedNotices] });
+  }
+
+  updateNotificationsByKind(notices: Notice[], kind: NoticeKinds) {
+    switch (kind) {
+      case 'access-lists':
+        const filtered = this.state.notices.filter(
+          n => n.kind !== 'access-lists'
+        );
+        this.setNotifications([...filtered, ...notices]);
+    }
+  }
```
sorry i should've mentioned that this is used in the enterprise: https://github.com/gravitational/teleport.e/pull/2373/files#diff-687e0165d14e46fad5b37841d13e11e7b38effc63892189bdbf19932b9ce7f9aR60, referencing a older code, still need to update the changes from this PR into that PR

kimlisa committed:
Address review

- Made store notification type more generic
- Drop opactiy styling for top bar drop downs
- Use assertUnreachable in switch cases

kimlisa commented:
> FYI we already have `Notifications` in `shared` (you can find them in storybook). We are using them in Connect and Isaiah used them [for a desktop session warnings](http://localhost:9002/?path=/story/teleport-desktopsession--warnings). I'm wondering if these things could (and should) be merged. I also know that our design team is working on a [design system for notifications](https://www.figma.com/file/Gpjs9vjhzUKF1GDbeG9JGE/Application-Design-System?type=design&node-id=8890-5502&mode=design&t=STFnnxzxKYsQPFJ3-0).
> 
> Probably these things are different enough that they should be separated (notifications from `shared` are more like alerts, while your notifications are more like an inbox). I'm only wondering if there is a long term vision in this - for example, you when you get a notification, you first see it as an alert, and then it goes to that _inbox_.

this seems like a xin/kenny question, but likely it will evolve in the future as we support more types of notification.

gzdunek commented on web/packages/teleport/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -67,7 +67,6 @@ export const DropdownItem = styled.div`
 `;
 
 export const commonDropdownItemStyles = css`
-  opacity: 0.8;
```
We should remove the `transition` and `hover` properties below as well.

gzdunek commented on web/packages/teleterm/src/ui/utils/assertUnreachable.ts:
```diff
@@ -14,6 +14,9 @@
  * limitations under the License.
  */
 
+/**
+ * @deprecated Import assertUnreachable from `shared/utils/error` instead.
+ */
 export function assertUnreachable(x: never): never {
   throw new Error(`Unhandled case: ${x}`);
 }
```
Thanks for this. 
Small improvement: we can just re-export the function from `shared` here, instead of declaring it again.

gzdunek commented on web/packages/shared/utils/error.ts:
```diff
@@ -30,3 +30,7 @@ export function getErrMessage(err: unknown) {
   return message;
 }
+
+export function assertUnreachable(x: never): never {
+  throw new Error(`Unhandled case: ${x}`);
+}
```
I think we can create a new file for this. It is not a function that checks error types.

gzdunek commented on web/packages/teleport/src/stores/storeNotifications.ts:
```diff
@@ -0,0 +1,82 @@
+};
+
+export type NotificationState = {
+  notices: Notification[];
```
Nit: `notifications`?

kimlisa committed:
Address CR

gzdunek reviewed (approved):
With some minor comments.

gzdunek commented on web/packages/teleport/src/Discover/Database/CreateDatabase/useCreateDatabase.ts:
```diff
@@ -16,7 +16,7 @@
 
 import useAttempt from 'shared/hooks/useAttemptNext';
-import { getErrMessage } from 'shared/utils/errorType';
+import { getErrMessage } from '@gravitational/shared/utils/errorType';
```
The import should be from `shared/utils/errorType`, not `@gravitational/shared/utils/errorType`.
It should be fixed in other changed files too.

kimlisa replied:
oops auto correct

kimlisa committed:
Fix import path and rename file

kimlisa committed:
Leave a TODO comment

kimlisa committed:
Merge branch 'master' into lisa/web/notification-bell
